backend/pkg/api/connect/service/clusterstatus/service.go

```diff
@@ -0,0 +1,332 @@
+	v1alpha2 "github.com/redpanda-data/console/backend/pkg/protogen/redpanda/api/dataplane/v1alpha2"
+	"github.com/redpanda-data/console/backend/pkg/version"
+)
+
+var _ consolev1alpha1connect.ClusterStatusServiceHandler = (*Service)(nil)
+
+// Service that implements the UserServiceHandler interface. This includes all
+// RPCs to retrieve cluster information about all connected APIs.
+type Service struct {
+	cfg    *config.Config
+	logger *zap.Logger
+
+	kafkaClientProvider    kafkafactory.ClientFactory
+	redpandaClientProvider redpanda.ClientFactory
+	schemaClientProvider   schema.ClientFactory
+	connectSvc             *kafkaconnect.Service
+
+	kafkaStatusChecker    *kafkaStatusChecker
+	redpandaStatusChecker *redpandaStatusChecker
+}
+
+// NewService creates a new Service that serves the RPCs for retrieving cluster statuses.
+func NewService(
+	cfg *config.Config,
+	logger *zap.Logger,
+	kafkaClientProvider kafkafactory.ClientFactory,
+	redpandaClientProvider redpanda.ClientFactory,
+	schemaClientProvider schema.ClientFactory,
+	connectSvc *kafkaconnect.Service,
+) *Service {
+	return &Service{
+		cfg:    cfg,
+		logger: logger,
+
+		kafkaClientProvider:    kafkaClientProvider,
+		redpandaClientProvider: redpandaClientProvider,
+		schemaClientProvider:   schemaClientProvider,
+		connectSvc:             connectSvc,
+
+		kafkaStatusChecker: &kafkaStatusChecker{logger: logger},
+	}
+}
+
+// GetKafkaInfo retrieves Kafka cluster metadata and API version concurrently,
+// aggregates details (such as broker counts, topics, partitions, and replicas),
+// and returns a comprehensive Kafka status response.
+func (s *Service) GetKafkaInfo(ctx context.Context, _ *connect.Request[v1alpha1.GetKafkaInfoRequest]) (*connect.Response[v1alpha1.GetKafkaInfoResponse], error) {
+	_, adminCl, err := s.kafkaClientProvider.GetKafkaClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// We use a child context with a shorter timeout because otherwise we'll potentially have very long response
+	// times in case of a single broker being down.
+	childCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	grp, grpCtx := errgroup.WithContext(childCtx)
+
+	// Fetch cluster metadata
+	var metadata kadm.Metadata
+	grp.Go(func() error {
+		var err error
+		metadata, err = adminCl.Metadata(grpCtx)
+		return err
+	})
+
+	// Fetch Kafka API version
+	clusterVersion := "unknown"
+	grp.Go(func() error {
+		var err error
+
+		apiVersions, err := adminCl.ApiVersions(ctx)
+		if err != nil {
+			s.logger.Warn("failed to request kafka version", zap.Error(err))
+		}
+		apiVersions.Each(func(versions kadm.BrokerApiVersions) {
+			if versions.Err != nil {
+				s.logger.Warn("failed to request kafka version", zap.Int32("broker_id", versions.NodeID), zap.Error(versions.Err))
+				return
+			}
+			clusterVersion = versions.VersionGuess()
+		})
+
+		return nil
+	})
+
+	if err := grp.Wait(); err != nil {
+		return nil, apierrors.NewConnectError(
+			connect.CodeInternal,
```

Comment:
Probably not a big deal...
Only the metadata call returns an error from the 2 goroutines... 
Would we be able to map that error to a more appropriate error perhaps?

Almost feels like maybe we need something like `NewConnectErrorFromKafkaErrorCode` but it accepts an error, and checks the rest (is unauthorized first, and then returns internal).

Reply:
That is intentional. We always want these cluster status rpcs to respond, even if the cluster is broken. If we would return a connectrpc error we cannot really inform about the perceived cluster status. If the metadata request fails, we cannot even figure out the health of the cluster and it has to be assumed lost connection, so that's the only case in my mind that makes sense to return an error.